DECISIONS.md

```diff
@@ -0,0 +1,119 @@
+  - make the renamed column `nullable`
+  - create a new column of name `original_field_name__b`
+  - stream new data to `original_field_name__b`
+  - (to see a full list of type suffixes, please see: [`json_schema._shorthand_mappings`](https://github.com/datamill-co/target-postgres/blob/master/target_postgres/json_schema.py#L283))
+
+#### Why
+
+***TL;DR:*** Instead of throwing a hard error and forcing users to do some manual
+transformation _before_ streaming data through `target-sql`, we chose a "best
+effort" approach to resolving the underlying error.
+
+By renaming and migrating the column we:
+
+- make the resulting structure in the database the same no matter whether we upload column `integer` _then_ column `boolean` or vice versa.
+- users learn of dependent views/columns blocking a type change _early_
+
+### Column Name Collision
+
+#### What
+
+- When attempting to `upsert_table`, `SQLInterface` has to handle name
+
+#### Why
+
+***TL;DR:*** Instead of throwing a hard error and forcing users to do some manual
+transformation _before_ streaming data through `target-sql`, we chose a "best
+effort" approach to resolving the underlying error.
+
+
+
+### Column Name Length
+
+#### What
+
+- `SQLInterface` provides a single field called `IDENTIFIER_FIELD_LENGTH` which is to be overridden by the implementing class
+- Any column which is found to be excess of `IDENTIFIER_FIELD_LENGTH` is truncated to be no longer than `IDENTIFIER_FIELD_LENGTH`
+- All `collision` and `type` information is preserved in the truncation
+  - ie, any values which are suffixed onto the name as `__...`
+- All original field/column names are preserved as a `column_mapping`
+
+#### Why
+
+***TL;DR:*** Instead of throwing a hard error and forcing users to do some manual
+transformation _before_ streaming data through `target-sql`, we chose a "best
+effort" approach to resolving the underlying error.
+
+Most (all?) SQL targets we have encountered have length restrictions for identifiers
+in their schema. Since arbitrary JSON _does_ not have this same restriction, we needed
+a best effort mechanism for handling names which were either auto-generated and are
+too long, or user input fields which physically cannot fit into the remote target.
+
+As such, we chose to take the simplest method here for clarity. ie, truncate the
+original/generated name, and then proceed with collision support as normal.
+
+The implementing class is tasked with providing `canonicalize_identifier`, a method
+which when called is expected to _only_ transform a string identifier into another
+string identifier which contains only characters which are allowed by the remote target.
+
+## Data De-nesting
+
+### Objects
+
+#### What
+
+- `Objects` are unpacked into their parent table.
+- The unpacked fields are prefixed with the name of the `field` which originally contained the object.
+
+#### Why
+
+This approach is inspired by what Stitch Data takes with `object` de-nesting.
```

Comment:
More of a root why, accessing nested objects as first class fields provides a better user experience.